torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,287 @@
+            output = self.norm(output)
+
+        return output
+
+
+class TransformerDecoder(Module):
+    r"""TransformerDecoder is a stack of N decoder layers
+
+    Args:
+        decoder_layer: an instance of the TransformerDecoderLayer() class (required). 
+        num_layers: the number of sub-decoder-layers in the decoder (required).
+        norm: the layer normalization component (optional).
+
+    Examples::
+        >>> transformer_decoder = nn.TransformerDecoder(decoder_layer, num_layers)
+    """
+
+    def __init__(self, decoder_layer, num_layers, norm=None):
+        super(TransformerDecoder, self).__init__()
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self.num_layers = num_layers
+        self.norm = norm
+
+    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+        r"""Pass the inputs (and mask) through the decoder layer in turn.
+
+        Args:
+            tgt: the sequence to the decoder (required).
+            memory: the sequnce from the last layer of the encoder (required).
+            tgt_mask: the mask for the tgt sequence (optional).
+            memory_mask: the mask for the memory sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        output = tgt
+
+        for i in range(self.num_layers):
+            output = self.layers[i](output, memory, tgt_mask, memory_mask)
+
+        if self.norm:
+            output = self.norm(output)
+
+        return output
+
+class TransformerEncoderLayer(Module):
+    r"""TransformerEncoderLayer is made up of self-attn and feedforward network.
+
+    Args:
+        d_model: the number of expected features in the input (required).
+        nhead: the number of heads in the multiheadattention models (required).
+        dim_feedforward: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> encoder_layer = nn.TransformerEncoderLayer(d_model, nhead)
+    """
+
+    def __init__(self, d_model, nhead, dim_feedforward=2048, dropout=0.1):
+        super(TransformerEncoderLayer, self).__init__()
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        # Implementation of Feedforward model
+        self.linear1 = Linear(d_model, dim_feedforward)
+        self.dropout = Dropout(dropout)
+        self.linear2 = Linear(dim_feedforward, d_model)
+
+        self.norm1 = LayerNorm(d_model)
+        self.norm2 = LayerNorm(d_model)
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+
+    def forward(self, src, src_mask=None):
+        r"""Pass the input through the endocder layer.
+
+        Args:
+            src: the sequnce to the encoder layer (required).
+            src_mask: the mask for the src sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        src2 = self.norm1(src)
```

Comment:
You should note that there's some debate about whether these go.

Reply:
One option is to check before calling the norm function.

Reply:
@stephenroller - do you mean there's debate around what kind of normalization to use or any at all?

@zhangguanheng66 - the docs should mention that we use norm, but I don't think we should make that optional / replaceable for now (unless we get many more comments).

Reply:
It might be easier to reimplement the EncoderLayer than making this one more complicated. But maybe I'm wrong.

Reply:
We should finally write a post to review the nn.Transformer module and clearly lay out the architecture, including norm.

Reply:
sorry I wrote whether but I meant "where".

Some variations put them before the selfattn/relu. Some put them after.

Reply:
Maybe we should benchmark a bunch of alternatives and pick the best one?

Reply:
Myle and Guillaume have for a bunch of things already. I’ve found it makes a *small* difference in my experiments but usually had too many confounding variables to make a clear recommendation.

Reply:
I suggest we do that in a separate PR since it's not a necessary feature to get the basic featureset.

Reply:
Just to actually ground it in code:

https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L532-L557

Reply:
Ok, let's standardize on ["Attention is all you need"](https://arxiv.org/abs/1706.03762) for now and mention that in the docs. Later on we can extend the behavior to match the implementation of other Transformer implementations with corresponding paper references.

Reply:
Also see [a comment in fairseq mentioning this](https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L570).